lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -0,0 +1,415 @@
+	I::read(&mut s.to_slice())
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// A Rust str object, ie a reference to a UTF8-valid string.
+/// This is *not* null-terminated so cannot be used directly as a C string!
+pub struct Str {
+	pub chars: *const u8,
+	pub len: usize
+}
+impl Into<Str> for &'static str {
+	fn into(self) -> Str {
+		Str { chars: self.as_ptr(), len: self.len() }
+	}
+}
+impl Into<&'static str> for Str {
+	fn into(self) -> &'static str {
+		if self.len == 0 { return ""; }
+		std::str::from_utf8(unsafe { std::slice::from_raw_parts(self.chars, self.len) }).unwrap()
+	}
+}
+
+// Note that the C++ headers memset(0) all the Templ types to avoid deallocation!
+// Thus, they must gracefully handle being completely null in _free.
+
+/// CSliceTempl may have the word slice in it, but its really a Vec. Because its used for things
+/// that require per-item conversion, we have to use a new buffer for it. Of course for opaque
+/// types, each item will simply be a pointer to the original slice's corresponding item, so we may
+/// not own the actual Rust item, only the buffer.
+///
+/// Thus, the implementation here is basically just that of CVecTempl.
+#[repr(C)]
+pub struct CSliceTempl<T> {
+	pub data: *mut T,
+	pub datalen: usize
+}
+impl<T> CSliceTempl<T> {
+	pub(crate) fn into_rust_vec(mut self) -> Vec<T> {
+		if self.datalen == 0 { return Vec::new(); }
+		let ret = unsafe { Box::from_raw(std::slice::from_raw_parts_mut(self.data, self.datalen)) }.into();
+		self.data = std::ptr::null_mut();
+		self.datalen = 0;
+		ret
+	}
+}
+impl<T> Drop for CSliceTempl<T> {
+	fn drop(&mut self) {
+		if self.datalen == 0 { return; }
+		unsafe { Box::from_raw(std::slice::from_raw_parts_mut(self.data, self.datalen)) };
+	}
+}
+
+// TODO: Integer/bool primitives should avoid the pointer indirection for underlying types
+// everywhere in the containers.
+
+#[repr(C)]
+pub union CResultPtr<O, E> {
+	pub result: *mut O,
+	pub err: *mut E,
+}
+#[repr(C)]
+pub struct CResultTempl<O, E> {
+	pub contents: CResultPtr<O, E>,
+	pub result_good: bool,
+}
+impl<O, E> CResultTempl<O, E> {
+	pub(crate) extern "C" fn good(o: O) -> Self {
+		CResultTempl {
+			contents: CResultPtr {
+				result: Box::into_raw(Box::new(o)),
+			},
+			result_good: true,
+		}
+	}
+	pub(crate) extern "C" fn err(e: E) -> Self {
+		CResultTempl {
+			contents: CResultPtr {
+				err: Box::into_raw(Box::new(e)),
+			},
+			result_good: false,
+		}
+	}
+}
+pub extern "C" fn CResultTempl_free<O, E>(_res: CResultTempl<O, E>) { }
+impl<O, E> Drop for CResultTempl<O, E> {
+	fn drop(&mut self) {
+		if self.result_good {
+			if unsafe { !self.contents.result.is_null() } {
+				unsafe { Box::from_raw(self.contents.result) };
+			}
+		} else if unsafe { !self.contents.err.is_null() } {
+			unsafe { Box::from_raw(self.contents.err) };
+		}
+	}
+}
+
+#[repr(C)]
+pub struct CVecTempl<T> {
```

Comment:
I did not understand the difference between this and `CSliceTempl`

Reply:
Neither do I! Its on my todo list to unify them :)

Reply:
This turned out to be a lot of work, but I did eventually finish this :).